File: web/src/core/protocol.ts
// 与 Rust 信令服务器 (src/protocol.rs) 一一对应的线格式。
//
// 服务端使用 #[serde(tag = "t", rename_all = "kebab-case",
// rename_all_fields = "camelCase")]，因此：
//   - 变体标签在字段 "t"，kebab-case（如 "peer-join"）
//   - 各变体内部字段为 camelCase（如 peerId）

/** 房间成员信息。 */
export interface PeerInfo {
  peerId: string
  nick?: string | null
}

/** ICE 服务器条目（STUN 无凭证；TURN 带临时凭证）。 */
export interface IceServer {
  urls: string[]
  username?: string | null
  credential?: string | null
}

/**
 * 内置 STUN/TURN 的接入信息：服务端只知道端口与凭证，
 * 主机名由前端取 location.hostname 拼接（客户端能开网页 ⇒ 该主机可达）。
 */
export interface BuiltinIce {
  /** 内置 STUN/TURN 的 UDP 端口；0 表示未启用。 */
  udpPort: number
  /** 内置 TURN over TCP 端口；0 表示未启用。 */
  tcpPort: number
  username: string
  credential: string
}

/** 客户端 → 服务端。 */
export type ClientMsg =
  | { t: 'join'; room: string; peerId: string; nick?: string | null }
  | { t: 'leave' }
  | { t: 'signal'; to: string; data: SignalData }
  | { t: 'turn-creds' }

/** 服务端 → 客户端。 */
export type ServerMsg =
  | { t: 'joined'; peerId: string; peers: PeerInfo[] }
  | { t: 'peer-join'; peer: PeerInfo }
  | { t: 'peer-left'; peerId: string }
  | { t: 'signal'; from: string; data: SignalData }
  | { t: 'turn-creds'; iceServers: IceServer[]; ttl: number; builtin?: BuiltinIce | null }
  | { t: 'error'; code: string; msg: string }

/**
 * 端到端信令负载：服务端仅透传，不解析。
 * 前两种取自 MDN「完美协商」示例；relayKey 是降级到 WS 中继时交换的
 * X25519 公钥（base64）——收到它同时意味着「对端已降级」，本端应一并切到中继。
 */
export type SignalData =
  | { description: RTCSessionDescriptionInit; candidate?: undefined; relayKey?: undefined }
  | { candidate: RTCIceCandidateInit | null; description?: undefined; relayKey?: undefined }
  | { relayKey: string; description?: undefined; candidate?: undefined }

File: src/protocol.rs
//! 信令消息协议（JSON over WebSocket）。
//!
//! 服务器只负责：房间成员管理、SDP/ICE 候选中继、TURN 临时凭证签发。
//! `signal` 消息里的 `data` 对服务器是不透明的（SDP 或 ICE candidate），
//! 服务器原样转发，绝不解析业务含义。

use serde::{Deserialize, Serialize};

/// 房间成员的公开信息。
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PeerInfo {
    pub peer_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nick: Option<String>,
}

/// 一个 ICE 服务器条目（STUN 或 TURN），下发给前端用于 `RTCPeerConnection`。
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IceServer {
    pub urls: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential: Option<String>,
}

/// 客户端 → 服务器。
#[derive(Debug, Deserialize)]
#[serde(tag = "t", rename_all = "kebab-case", rename_all_fields = "camelCase")]
pub enum ClientMsg {
    /// 加入房间；由后到者随后发起 offer，避免双向 glare。
    Join {
        room: String,
        peer_id: String,
        nick: Option<String>,
    },
    /// 主动离开房间。
    Leave,
    /// 向房间内某个对端转发信令（SDP / ICE candidate）。
    Signal {
        to: String,
        data: serde_json::Value,
    },
    /// 请求 TURN/STUN 临时凭证。
    TurnCreds,
}

/// 服务器 → 客户端。
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "t", rename_all = "kebab-case", rename_all_fields = "camelCase")]
pub enum ServerMsg {
    /// 加入成功，附带房间内已有成员列表。
    Joined {
        peer_id: String,
        peers: Vec<PeerInfo>,
    },
    /// 有新成员加入。
    PeerJoin { peer: PeerInfo },
    /// 有成员离开。
    PeerLeft { peer_id: String },
    /// 收到对端转发来的信令。
    Signal {
        from: String,
        data: serde_json::Value,
    },
    /// TURN/STUN 凭证响应；`ttl` 为凭证有效期（秒）。
    TurnCreds {
        ice_servers: Vec<IceServer>,
        ttl: u64,
    },
    /// 错误。
    Error { code: String, msg: String },
}

impl ServerMsg {
    pub fn error(code: &str, msg: &str) -> Self {
        ServerMsg::Error {
            code: code.to_string(),
            msg: msg.to_string(),
        }
    }
}
